Using the new 0.9 Kafka consumer.

alexlod opened the pull request:
Please review carefully as this is my first schema registry patch. Plus, I'm not intimately familiar with the old consumer and may not have replicated its semantics perfectly. Although, I think I did. Thanks!

ewencp commented:
@granders Can you take a pass at reviewing this?

hachikuji commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStore.java:
```diff
@@ -106,12 +107,19 @@ public KafkaStore(SchemaRegistryConfig config,
-                                         serializer, this.localStore, this.noopKey);
     this.brokerSeq = zkUtils.getAllBrokersInCluster();
 
+    List<Broker> brokers = JavaConversions.seqAsJavaList(brokerSeq);
```
I would have thought we'd rather populate the bootstrap brokers from configuration than from zookeeper.

ewencp replied:
@hachikuji We need to figure out the compatibility story on that. We would prefer that, but we used to be on the old consumer and we need ZK info since schema registry uses it to do it's own leadership election. Since we already have the ZK config, it probably doesn't make sense to provide a separate broker list unless we have a path for getting away from ZK entirely

hachikuji replied:
Figured that was the reason. It's probably fine since the broker nodes should always be world readable.

hachikuji commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -82,54 +77,53 @@ public KafkaStoreReaderThread(ZkUtils zkUtils,
+    this.consumer.assign(Arrays.asList(this.topicPartition));
+    this.consumer.poll(1); // poll to force communication with the broker before checking if the assignment exists.
+    this.consumer.seekToBeginning(this.topicPartition);
+    if (this.consumer.assignment().size() != 1) {
```
nitpick: we can probably give a better error message below if size() is greater than 1.

hachikuji replied:
On second thought, this is actually kind of weird. We assign the single partition explicitly above, so there's no reason to check if the assignment is different. Perhaps you should call partitionsFor() and verify that there is exactly one partition. You can skip the poll(1) as well.

alexlod replied:
Good suggestions. Also, good call removing poll(1). Would only need that if I used subscribe() (AFAIK).

hachikuji commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -171,11 +165,14 @@ public void doWork() {
-      throw new IllegalStateException(
-          "KafkaStoreReaderThread's ConsumerIterator timed out despite expected infinite timeout.");
+    } catch (WakeupException we) {
+      this.consumer.close();
```
Maybe this should go in a finally block?

alexlod replied:
Ah, good suggestion. Originally I thought that would cause the consumer to close prematurely. But that isn't true because I'm using Long.MAX_VALUE in poll().

alexlod replied:
Adding close() to the finally block caused "consumer already closed" exceptions. In my recent commit, I kept the close() call in the catch statement. However, I'm confused by this and will investigate it.

hachikuji replied:
It's probably because you also have a close() that follows the call to wakeup(). If you have close() in the finally block, then maybe you can get rid of the other call.

hachikuji commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -171,11 +165,14 @@ public void doWork() {
-          "KafkaStoreReaderThread's ConsumerIterator timed out despite expected infinite timeout.");
+    } catch (WakeupException we) {
+      this.consumer.close();
     } catch (MessageSizeTooLargeException mstle) {
```
Probably need to replace this with the client variant: RecordTooLargeException.

alexlod replied:
Yes, and update the error message.

hachikuji commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -171,11 +165,14 @@ public void doWork() {
           }
         }
+
+        if (commitInterval > 0 && System.currentTimeMillis() - lastCommitTime > commitInterval) {
```
Missing an update to `lastCommitTime`?

alexlod replied:
Good catch! That was missing before this patch, too.

granders replied:
Looking back on this, I think we want to a) remove commit-related logic entirely, or b) actually design a snapshot mechanism that makes use of commits.

I don't see (b) as happening, or as particularly necessary, so probably we want to remove commit logic entirely.

alexlod replied:
Will do!

hachikuji commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -208,7 +200,9 @@ public void waitUntilOffset(long offset, long timeout, TimeUnit timeUnit) throws
     }
-    
+
+    log.trace("Waiting for offset to be " + offset + ". Currently " + offsetInSchemasTopic);
```
Nitpick: This is a little nicer:

```java
log.trace("Waiting for offset to be {}. Currently {}", offset, offsetInSchemasTopic);
```
Also, the message is a little unclear. Maybe something like "Waiting to read to offset {}"?

alexlod replied:
+1

hachikuji commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStore.java:
```diff
@@ -106,12 +107,19 @@ public KafkaStore(SchemaRegistryConfig config,
 
+    List<Broker> brokers = JavaConversions.seqAsJavaList(brokerSeq);
+    StringBuilder sb = new StringBuilder();
+    for (int i = 0; i < brokers.size(); i++) {
```
Do you need the index here?

alexlod replied:
Nope. Will change to a for each. Good catch.

hachikuji commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStore.java:
```diff
@@ -146,8 +143,13 @@ public void init() throws StoreInitializationException {
+    // the thread must be created after the schema topic has been created.
+    this.kafkaTopicReader =
+            new KafkaStoreReaderThread<>(this.bootstrapBrokers, topic, groupId,
+                    Integer.MIN_VALUE, this.storeUpdateHandler,
```
Not part of this patch, but I'm kind of puzzled with commit interval passed here. Doesn't this effectively turn off commits? If so, then do we even need the commit logic implemented in KafkaStoreReaderThread?

granders replied:
It may have been an oversight that we ended up even exposing `KAFKASTORE_COMMIT_INTERVAL_MS_CONFIG` since commits are not used at all.

Unless we implement some sort of snapshot mechanism or use a backing store with that capability, it's probably better to get rid of this config altogether since it does nothing. Which means we continue to do what we already do - rebuild the in-memory caches by consuming from offset 0 if we restart the process. 

I think the reasoning here is that the number of schemas in the lifetime of an organization is small enough that building from t=0 on startup is ok.

alexlod commented:
Adding a new commit addressing some of @hachikuji's suggestions. Thanks!

granders commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStore.java:
```diff
@@ -106,12 +107,19 @@ public KafkaStore(SchemaRegistryConfig config,
+    List<Broker> brokers = JavaConversions.seqAsJavaList(brokerSeq);
+    StringBuilder sb = new StringBuilder();
+    for (Broker broker : brokers) {
+      for(EndPoint ep : JavaConversions.asJavaCollection(broker.endPoints().values())) {
```
Presumably this patch is a step in the direction of enabling security (yay!). But while we still really only support plaintext, we may need to think about the fact that collecting every endpoint, as in this block of code, would cause us to include non-plaintext endpoints in our list.

alexlod replied:
Cool. I'll throw an exception if a non-plaintext endpoint is found.

ewencp replied:
Might not want to throw an exception -- some folks might want to use both secure and plaintext listeners, even if not everything has migrated over to secure access yet. Throwing an exception would break this for them without a way to override it. It might be better to filter out secure endpoints, log a warning, and throw an exception if you end up with an empty list.

alexlod replied:
Actually, I think a better approach would be to `log.warn` when a non-plaintext is found, and ignore it. And throw an exception if there are no plaintext endpoints.

granders commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -82,54 +77,58 @@ public KafkaStoreReaderThread(ZkUtils zkUtils,
-    consumerProps.put("zookeeper.connect", kafkaClusterZkUrl);
-    consumerProps.put("auto.offset.reset", "smallest");
+
+    consumerProps.put("bootstrap.servers", bootstrapBrokers);
```
Would it make sense to import `ConsumerConfig` and use for example `BOOTSTRAP_SERVERS_CONFIG` instead of `"boostrap.servers"`. 

Similar comment for other configs.

alexlod replied:
Good call. Will do.

granders commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -82,54 +77,58 @@ public KafkaStoreReaderThread(ZkUtils zkUtils,
-    Option<String> committedOffsetStringOpt = zkUtils.readDataMaybeNull(
-        String.format("/consumers/%s/offsets/%s/0", group, topic))._1();
-    if (committedOffsetStringOpt.isEmpty()) {
+  private long offsetOfLastConsumedMessage() {
```
If we ditch commit logic as mentioned in other comments, we don't need this method.

alexlod replied:
Good catch.

granders commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -171,31 +170,30 @@ public void doWork() {
     log.debug("Starting shutdown of KafkaStoreReaderThread.");
     if (consumer != null) {
-      consumer.shutdown();
+      consumer.wakeup();
```
Do we want to add a countdown latch as outlined in the "shutdown with wakeup" example?
http://docs.confluent.io/2.0.1/clients/consumer.html#shutdown-with-wakeup

alexlod replied:
I don't think we need it because `ShutdownableThread` (which `KafkaStoreReaderThread` extends) uses a countdown latch. Although, this does mean I need to change the shutdown logic to not just call `super.shutdown()`.

granders commented on core/src/test/java/io/confluent/kafka/schemaregistry/client/CachedSchemaRegistryClientTest.java:
```diff
@@ -64,41 +58,37 @@ private IndexedRecord createAvroRecord() {
-    props.put("zookeeper.sync.time.ms", "200");
+    props.put("session.timeout.ms", "6000"); // default value of group.min.session.timeout.ms.
+    props.put("heartbeat.interval.ms", "2000");
     props.put("auto.commit.interval.ms", "1000");
```
Can delete this assuming we ditch commit logic

granders commented:
@alexlod Thanks for pinging me, sorry I lost track of this. My main observation revolves around removing offset commit logic which was actually not used at all before.

@ewencp Second opinion on ditching commit-related logic? Basically I think we always want to consume from t=0

hachikuji commented:
Where did we land on ripping out the commit logic in KafkaStoreReaderThread? Seems like it makes sense given current usage.

alexlod commented:
@granders: great suggestions. Thanks! I addressed all of them in my most recent commit.

@hachikuji: I just ripped it out in my most recent commit.

hachikuji commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -79,57 +71,49 @@ public KafkaStoreReaderThread(ZkUtils zkUtils,
+            org.apache.kafka.common.serialization.ByteArrayDeserializer.class);
+    this.consumer = new KafkaConsumer<>(consumerProps);
+
+    int partitionCount = this.consumer.partitionsFor(this.topic).size();
```
It looks like partitionsFor() returns null instead of an empty collection when the topic is unknown (I should probably clarify that in the KafkaConsumer docs). It might be safest to check both for null and size() < 1.

alexlod replied:
Great catch. Thanks!

hachikuji commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -79,57 +71,49 @@ public KafkaStoreReaderThread(ZkUtils zkUtils,
-      if (consumerIterator.hasNext()) {
-        MessageAndMetadata<byte[], byte[]> messageAndMetadata = consumerIterator.next();
-        byte[] messageBytes = messageAndMetadata.message();
+      // don't need to commit offsets because the log is always read from the beginning on startup
```
Minor: this is sort of a strange location to see a comment like this. I'd consider either removing it or possibly moving it to the top of the class with a fuller description of the implementation.

alexlod replied:
Good suggestion. I wasn't sure where to put it, either. I like the idea of putting it at the top of the class.

hachikuji commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -147,7 +131,7 @@ public void doWork() {
-                messageBytes == null ? null : serializer.deserializeValue(messageKey, messageBytes);
+                record.value() == null ? null : serializer.deserializeValue(messageKey, record.value());
           } catch (SerializationException e) {
             log.error("Failed to deserialize a schema or config update", e);
```
Not part of this patch, but it seems like a bug to ignore this exception and continue below. Wouldn't we then incorrectly see the message as an instance of null and delete the key?

alexlod replied:
I'll throw an exception here. I'll do the same for the other occurrence you commented on, and an additional occurrence (deserializing the key). Great suggestion.

hachikuji commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -79,57 +71,49 @@ public KafkaStoreReaderThread(ZkUtils zkUtils,
-          messageKey = this.serializer.deserializeKey(messageAndMetadata.key());
+          messageKey = this.serializer.deserializeKey(record.key());
         } catch (SerializationException e) {
           log.error("Failed to deserialize the schema or config key", e);
```
Similar to the comment below. If we ignore this exception, I think we'd get an NPE on line 121.

alexlod commented:
New commit that addresses @hachikuji's suggestions. Thanks!

hachikuji commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -147,9 +141,9 @@ public void doWork() {
+                record.value() == null ? null : serializer.deserializeValue(messageKey, record.value());
           } catch (SerializationException e) {
-            log.error("Failed to deserialize a schema or config update", e);
+            throw new RuntimeException("Failed to deserialize a schema or config update", e);
```
Would it be better to just log and continue?

alexlod replied:
You commented on this previously:

> Not part of this patch, but it seems like a bug to ignore this exception and continue below. Wouldn't we then incorrectly see the message as an instance of null and delete the key?

Which I interpreted as "throw an exception instead." Did I misinterpret that?

hachikuji replied:
Sorry, what I meant is log and continue to the next message. The problem previously was that we caught the error and then proceeded with the logic below, which seems like it would incorrectly result in a deletion.

alexlod replied:
Ah right. Good catch!

hachikuji commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -162,45 +156,40 @@ public void doWork() {
-          "KafkaStoreReaderThread's ConsumerIterator timed out despite expected infinite timeout.");
-    } catch (MessageSizeTooLargeException mstle) {
+    } catch (WakeupException we) {
+      consumer.close();
```
Can you move this to a finally block? You can catch the WakeupException and just ignore it, which will let the finally block execute. I think we want to make sure that the consumer is closed in all cases. (I may have mentioned this before, so apologies if there's a good reason to keep it as it is.)

alexlod replied:
It can't just be moved to the finally block, because doWork() is called from a `while(true)` loop. So putting it in `finally` will prematurely close it. However, you're right that there could be a case where the consumer isn't closed. An alternative could be to use an AtomicBoolean, right?

hachikuji replied:
Ah, that makes sense. I forget we were using ShutdownableThread. In that case, I wonder if it makes sense to catch and ignore the WakeupException and then call close() in the shutdown() hook after the call to `super.shutdown()`. That ensures that the loop has actually exited before close() is called. Otherwise it seems we have a couple race conditions where it's possible that we either 1) continue using the consumer after it has been closed, or 2) never call close() at all.

alexlod replied:
Nice, thanks for walking me through that. I was envisioning a much more complicated solution.

hachikuji commented:
@alexlod Nice work on this patch! Seems like it's getting close. Would you mind rebasing your changes?

granders commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStore.java:
```diff
@@ -182,12 +174,40 @@ private void createSchemaTopic() throws StoreInitializationException {
     }
   }
 
+  private String getBrokerEndpointsFromZk(Seq<Broker> brokers) {
```
Might be good to have a unit test for this? 

Also, I think we've seen from some users when they have some difficulties with setup that it would be useful to log the list of endpoints collected from zookeeper.

alexlod replied:
I'm happy to write unit tests. Should I make the method `public` or `protected` to facilitate this? What's the best practice here?

granders commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -30,43 +38,38 @@
-import kafka.utils.ZkUtils;
-import scala.Option;
 
+/**
```
Thanks for the added doc!

granders commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStore.java:
```diff
@@ -98,20 +101,15 @@ public KafkaStore(SchemaRegistryConfig config,
-    
-    // TODO: Do not use the commit interval until the decision on the embedded store is done
-    int commitInterval = config.getInt(SchemaRegistryConfig.KAFKASTORE_COMMIT_INTERVAL_MS_CONFIG);
+
```
I noticed that these constants (starting line 60) are all unused
```
  private static final String CLIENT_ID = "schema-registry";
  private static final int SOCKET_BUFFER_SIZE = 4096;
  private static final long LATEST_OFFSET = -1;
  private static final int CONSUMER_ID = -1;
```

not due to your changes - but might be worth cleaning up anyway

alexlod replied:
Good catch!

alexlod commented:
@granders and @hachikuji: thanks for your comments! I addressed many of them in a new commit (that is rebased). Although some of your comments are still open above (I had follow-up questions).

alexlod commented:
@granders and @hachikuji: I addressed @hachikuji's comments and pushed a new (rebased) commit. Thanks for your help, guys.

alexlod commented:
FYI: I'm working on running muckrake tests. I'll report back with results.

hachikuji commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -172,43 +168,41 @@ public void doWork() {
-          "KafkaStoreReaderThread's ConsumerIterator timed out despite expected infinite timeout.");
-    } catch (MessageSizeTooLargeException mstle) {
+    } catch (WakeupException we) {
+      // do nothing.
```
nitpick: maybe mention that we do nothing because the thread is closing?

hachikuji commented:
@alexlod Left one very minor comment, but overall LGTM (assuming system tests pass).

alexlod commented:
@hachikuji, thanks! I addressed the comment.

Also, looks like the system tests passed: https://jenkins.confluent.io/job/system-test-confluent-platform-branch-builder/7/console

hachikuji commented:
@alexlod Great work! Maybe we can let @granders take a final look before merging.

granders commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStore.java:
```diff
@@ -190,6 +177,37 @@ private void createSchemaTopic() throws StoreInitializationException {
     }
   }
 
+  private String getBrokerEndpointsFromZk(Seq<Broker> brokers) {
```
- maybe update name since this isn't actually connecting to zookeeper?
- also it'd be good to add a unit test for this method; probably reasonable to make it package private so it's available to test.

granders commented:
@alexlod Good stuff, just one more comment on adding a unit test.

alexlod commented:
@granders, great suggestion. I just pushed a new commit with some unit tests (I also changed the method signature of `getBrokerEndpoints`. Take a look, and when you say the word, I'll rebase and re-push.

(Also, thanks @hachikuji for the simple testing idea)

alexlod commented:
I rebased into one commit and also resolved a conflict with master. I think this thing is ready to go!

Ishiihara commented:
@alexlod We make some changes on the config. Can we make sure that the documentation is updated as well. Maybe in another patch.

alexlod commented:
@Ishiihara I saw that in the merge conflict I resolved an hour ago or so. I'm pretty sure I resolved this correctly. Take a look at the diff for `SchemaRegistryConfig` and `config.rst`.

Or maybe I'm misinterpreting your comment? Are you suggesting that I'm making some config changes but I'm not documenting this? The config change I made was to remove a config parameter that was inactive. Seems like that would be better for release notes rather than documentation?

Ishiihara commented:
@alexlod Sorry about that. I somehow missed the change to the documentation. Thanks for working on the patch.

alexlod commented:
Ishiihara no worries. My pleasure -- it's been fun! Thanks for reviewing it!

granders commented:
@alexlod Thanks much for adding the unit tests. LGTM

alexlod commented:
Thanks @granders, @Ishiihara and @hachikuji for all your help with this one! I'll merge it now.
